### ISSUE TITLE
"Error: rename ... Access is denied" When Creating a Ticket

### ISSUE THREAD

**Issue by blumf** [MichaelMure/git-bug#1142]
Been testing this out. If I make a local git repo, everything seems fine, but if I clone a remote repo, and try to use it I get an error:

(in a Windows command window)
```
> git clone https://server/git/git_bug_test
> cd git_bug_test
> git-bug user create
Name [me]:
Email [me@fake.com]:
Avatar URL:

bbc5207b0d2ddb8b823d53052a3a77eafc996a03143c3bed99f5841c2438e921

> git-bug add -t "First Bug" -m "Bug number one!"
Error: rename E:\temp\git_bug_test\.git\objects\pack\tmp_obj_2131903459 E:\temp\git_bug_test\.git\objects\e6\9de29bb2d1d6434b8b29ae775ad8c2e48c5391: Access is denied.
```

I can manually edit/rename the mentioned files just fine, no locks or access limits, so maybe it something else throwing the error with a spurious message?

**Comment by miwig** [comment]
Also seeing this on Windows. Possibly related to #809 ? i.e. access is denied because Windows considers the file in use?

**Comment by sudoforge** [comment]
> Also seeing this on Windows. Possibly related to #809 ? i.e. access is denied because Windows considers the file in use?

This seems unrelated to #809, which is focused on the use of parallelization in tests and cleaning up some temporary files created during the tests.

---

Can you provide more information about your environment when encountering this issue? Do you have the repository open in multiple windows, for example? Are you using Command Prompt, PowerShell, Git Bash (cygwin), or WSL2?

**Comment by roteiro** [comment]
I see the same issue, when trying to add bugs to an existing git repo. 
Environment: Windows 10, Powershell 5.1.22621.3958

Steps to reproduce
```
mkdir test_repo
git init
New-Item .gitignore
git add .\.gitignore
git commit -m 'add: .gitginore'
git bug user create
git bug add -t 'testbug' -m 'description' 
 
```
Result: 
```
Error: rename C:\Users\<myusername>\test_repo\.git\objects\pack\tmp_obj_3021921222 C:\Users\<myusername>\test_repo\.git\objects\e6\9de29bb2d1d6434b8b29ae775ad8c2e48c5391: Access is denied. 
```
The reason is that the files in `objects` are write-protected since they should never be modified by design of git.

However, when i initialize an empty repo and add a bug first before doing any other commits, this seems to work fine:
```
mkdir test_repo2
cd test_repo2
git init
git bug user create
git bug add -t 'title' -m 'description'
>fb67815 created
New-Item .gitignore
git commit -m 'add: .gitginore'
git bug add -t 'title2' -m 'description2'
>1bbb20e created

**Comment by sudoforge** [comment]
> Is it okay to mess with files like this? I thought that this tool could be cleanly integrated into a git environment.
> I want to use this tool but if it is going to clash with git in some way, I'm not really sure if I should use it.

`git-bug` shouldn't ever attempt to touch objects that it didn't create. so if the "safety" you're thinking of is related to your working tree and/or normal git commits and objects, yes, git-bug is safe.

**Comment by sudoforge** [comment]
regarding this issue:

as of `v0.8.1`, the "create a new bug" workflow is pretty simple. it handles a bit of logic like getting and cleaning up the title and body, and then calls a `go-git` method to handle the actual object creation.

so, i think this is likely to be a go-git issue. i found this issue, which seems to be directly related: https://github.com/go-git/go-git/issues/55. unfortunately, it doesn't have a linked commit indicating it was closed.

it [appears](https://github.com/src-d/go-git/issues/1155) that a workaround might be to use "git bash", but that obviously isn't ideal if you're used to powershell and want to keep using it.

i don't have a windows machine (physical or virtual) to recreate this issue, so unfortunately, i can't really debug or iterate on go-git's internals to work out a fix at the moment.

**Comment by sudoforge** [comment]
@MichaelMure do you have any insight to add here? i think a viable workaround in the short term to unblock users would be to call the git command line for this (on windows only, as in, a separate func with a `//go:build windows` directive).

**Comment by JoacoG** [comment]
Thank you for your quick response !

> it https://github.com/src-d/go-git/issues/1155 that a workaround might be to use "git bash", but that obviously isn't ideal if you're used to powershell and want to keep using it.

I tried this but I get this error

```
[13192:0511/212303.491:ERROR:service_worker_storage.cc(2072)] Failed to delete the database: Database IO error
Error: rename C:\Users\jguti\suite-cupones\.git\objects\pack\tmp_obj_3190080821 C:\Users\jguti\suite-cupones\.git\objects\e6\9de29bb2d1d6434b8b29ae775ad8c2e48c5391: Access is denied.
```

This same error appears in PowerShell7

This line is new:

`[13192:0511/212303.491:ERROR:service_worker_storage.cc(2072)] Failed to delete the database: Database IO error`

It did not appear this first time...

> I have the same problem in Windows 11

...that I got the error, so that maybe there is a problem in my end unrelated to this issue.

```
i don't have a windows machine (physical or virtual) to recreate this issue, so unfortunately, i can't really debug or iterate on go-git's internals to work out a fix at the moment.
```

Maybe if it is a quick test and not too complicated I could do it for you? Dumb suggestion I guess.

**Comment by JoacoG** [comment]
> it'd likely involve running delve to debug code while modifying git-bug and go-git, recompiling, and re-running. not something i'd deem as "quick", necessarily :P

Not really my definition of "quick" :S

---

I did the test and it worked. Apparently, creating a first bug using WSL "unlocks" the creation of bugs the normal way. These are the details:

I installed WSL and the archlinux distro. I installed git-bug manually with AUR (I did a fix stated in the comments [here](https://aur.archlinux.org/packages/git-bug-bin) to be able to install it). I cloned my project repo from GitHub into the archlinux filesystem, tried to create a bug and it succeeded.

I also tried to create a bug directly in the cloned version of my repo in the windows filesystem, where I always work, and this succeeded too. I used WSL here.

Then I tried to create a bug using Git Bash and PowerShell7 directly and it worked.

I created a few more bugs and they were created without problems.

I tried both termui and webui.

I'm still getting this error...
`[6516:0511/232729.476:ERROR:service_worker_storage.cc(2072)] Failed to delete the database: Database IO error`
...so that I think that this is not related to the issue. Probably this is due to some "tweaks" I was doing in my machine.

I didn't do further tests because I really don't now how to use this tool properly. If you want me to make some specific tests I could try to do it.

**Comment by sudoforge** [comment]
> [6516:0511/232729.476:ERROR:service_worker_storage.cc(2072)] Failed to delete the database: Database IO error

sorry, can you clarify when you're seeing this? is this showing on the command line? in the browser console? somewhere else?

**Comment by sudoforge** [comment]
> Then I tried to create a bug using Git Bash and PowerShell7 directly and it worked.

i'm going to write this behavior off as some ✨ windows magic ✨. i won't forget to test this some time `$LATER` when i set up a windows VM.

**Comment by roteiro** [comment]
> it [appears](https://github.com/src-d/go-git/issues/1155) that a workaround might be to use "git bash", but that obviously isn't ideal if you're used to powershell and want to keep using it.

I can confirm that using git bash does not solve the issue on Windows 11. (Neither does using git-bug `termui` nor `webui` for creating a first bug but I assume this was expected)

If you have any more ideas for workarounds on Windows which do not involve setting up WSL I volunteer to try them.

**Comment by sudoforge** [comment]
> If you have any more ideas for workarounds on Windows which do not involve setting up WSL I volunteer to try them.

if you are unable or unwilling to install the WSL (which appears to be a valid workaround based on this thread), the best i can do for you is to dig up my windows key and get a VM set up for recreating and iteratively debugging this. i'm not familiar enough with windows to take a blind stab at debugging this on another platform.

i can't provide a timeline for doing this right now -- it might be this week, it might be next month, and it could be (but i'd be surprised) if it were even later.

if you are not comfortable with golang, this isn't likely to be something easy for you to take on -- but if you are, i'd welcome the support :)

**Comment by JoacoG** [comment]
> However, when i initialize an empty repo and add a bug first before doing any other commits, this seems to work fine:

Regarding this issue, I also want to say that I tried this and it worked

**Comment by sudoforge** [comment]
@JoacoG i'm not sure what the exact source of the various messages you're getting are, but they definitely are not git-bug. the service worker error is most likely from electron (and thus, vscode).

> > However, when i initialize an empty repo and add a bug first before doing any other commits, this seems to work fine:
>
> Regarding this issue, I also want to say that I tried this and it worked

yes -- the original comment on this issue states that initializing git-bug in a brand new repository did not result in the error.

**Comment by blumf** [comment]
> the original comment on this issue states that initializing git-bug in a brand new repository did not result in the error.

To be more specific, it was a locally created repo that was fine. A remote repo, even a newly created empty one, threw the error.

Wondering if it's running into an issue of Windows being more strict in what you can do with an open file. *nix allows you to rename an open file, Windows won't (this is one of the reasons Windows updates often require a full reboot). If that's the case, stepping through with a debugger and checking what has the file open (stale file handle waiting to be GC'd?) might give a hint even on a non-Windows install.

(thanks all for looking into this, I know what a pain it is debugging on a system you don't use)

**Comment by JoacoG** [comment]
I really don't know what's going on but I did other tests.

> I see the same issue, when trying to add bugs to an existing git repo.
> Environment: Windows 10, Powershell 5.1.22621.3958
> 
> Steps to reproduce
> ```
> mkdir test_repo
> git init
> New-Item .gitignore
> git add .\.gitignore
> git commit -m 'add: .gitginore'
> git bug user create
> git bug add -t 'testbug' -m 'description' 
> ```

I did something similar to this but I used touch instead of New-Item. I'm using PowerShell7. This allowed me to create a first bug

**(1)**
```
mkdir test_repo
cd test_repo
git init
touch .gitignore
git add .\.gitignore
git commit -m 'add: .gitginore'
git bug user new
          ...CREATE THE USER...
git bug user adopt [HASH]
git bug termui
          ...CREATE THE BUG...
```

I also tested with Git Bash and it worked.

I also tried to create an empty repo and clone it in my machine, Then I did **(1)** but beginning in the line below git init, and I was able to create a first bug too. 

Now, about further bugs creations. I tried to add a file manually (test.txt) in both cases, then add, commit and create a bug. In the first case (local repo) I was able to create the second bug, but in the second case (empty cloned repo from github) I got the error. NOTE: I then tried this a second time and in both cases I was able to create the bug, I don't know what I did differently in the empty repo case

I also tried to use New-Item in the local repo test, add, commit and create a third bug and this succeeded too.

I really don't have much technical knowledge here, so that I cannot analyze this thoroughly, I even don't know if this is helpful, I hope it is.

**Comment by sudoforge** [comment]
@blumf yes, some windows behavior like that is likely to be the root cause here, and if so, we'd find this deeper within `go-git` (when it's actually creating the pack and object). nothing grabbed my attention yesterday when i stepped through the `git-bug-bug-new` workflow, but i could have missed something, as i'm not intimately familiar with the library yet (my existing usage of go-git has left me somewhat frustrated by its divergence from git's own behaviors, and puzzled at why golang doesn't have a better libgit2 module).

i was able to locate the windows license i used a handful of years ago (yay password managers, even if you do forget the item name). i'm going to bump up the priority on this, since it's a blocking issue for windows users. i'll be looking into this issue over the next week and will hopefully come back to this issue with much more (useful) information (and a fix).

**Comment by rebb** [comment]
Same problem. Kinda prevents me from using this.

Using 0.10.2 on Windows 10.

Also it seems not to play well with Sublime Text as the text editor. It doesn't even output the "access denied" error message when using it. Just silently fails and shows "Empty Title Aborting" inside the TermUI.

Only when using notepad.exe as the editor does it actually show the "access denied" error.

**Comment by sudoforge** [comment]
> Also it seems not to play well with Sublime Text as the text editor. It doesn't even output the "access denied" error message when using it. Just silently fails and shows "Empty Title Aborting" inside the TermUI.

@rebb i believe your issue is that you have `EDITOR` set to `subl`, which is non-blocking by default (it returns immediately upon being opened). this causes the command to exit immediately, at which point the file **is** empty, because you haven't typed anything in and saved it yet.

`subl` has the `--wait` (or `-w`) flag to change this behavior:

```
  -w or --wait:           Wait for the files to be closed before returning
```

`git-bug` [finds your editor][0] using the following, in order of preference (first one wins):

- `GIT_EDITOR` (if set)
- `core.editor` in your git configuration chain, if set
- `VISUAL` (if set)
- `EDITOR` (if set)
- binaries in your `PATH`: `editor`, `nano`, `vim`, `vi`, `emacs`, `ed`

addmitedly, we could probably support a custom variable specific to `git-bug`, but in practice, you likely want the same `subl -w` set for `GIT_EDITOR` (or `core.editor` in your git configuration).

[0]: https://github.com/git-bug/git-bug/blob/9b06f224354f3cf745ad6bf6620c1b04ca51077f/repository/gogit.go#L314-L354

**Comment by rebb** [comment]
How would i go about attaching the -w / --wait part ?

I tried adding it in the .gitconfig and GIT_EDITOR, but it fails with `Error: Unable to start editor: exec: "sh": executable file not found in %PATH%`, while the entry without the -w or --wait works, but has the mentioned problem.

I assume even if that worked i would still run into the "access denied" error though ? Not sure.

I would definitely be fine with using any editor, as long as it works - since this is just about a short text input.

**Comment by sudoforge** [comment]
what did you set `core.editor` and/or `GIT_EDITOR` to? that error message isn't what i'd expect to see if it's set to `subl -w`. maybe an issue with how the parsing occurs on windows... i'd have to check this out.

are you using powershell? command prompt? the "git for windows" terminal? wsl?

**Comment by rebb** [comment]
I'm using Tabby.sh with the Cmd(clink) profile.

I at least go to the "access denied" error now using Sublime-Text, by using sublime_text.exe instead of subl.exe.

One thing i'm doing differently though is calling the git-bug.exe directly instead of using the "git bug <..>" approach, because i couldn't figure out how to get that working with just the exe from the Releases page :(

Would that potentially cause issues ?

**Comment by sudoforge** [comment]
> One thing i'm doing differently though is calling the git-bug.exe directly instead of using the "git bug <..>" approach, because i couldn't figure out how to get that working with just the exe from the Releases page :(
>
> Would that potentially cause issues ?

no, calling the binary directly wouldn't cause any issues. as an fyi, the way you get the subcommand working is by moving the binary to somewhere in your `PATH`. when you invoke `git foo`, git first checks if `foo` is a valid internal subcommand, and if it isn't, it searches your `PATH` to see if a `git-foo` binary exists (in the case of windows, it checks for `git-foo.exe`).

---

i had some time and took another look today. the error you received:

```
Error: Unable to start editor: exec: "sh": executable file not found in %PATH%
```

actually comes from `git-bug`: https://github.com/git-bug/git-bug/blob/9b06f224354f3cf745ad6bf6620c1b04ca51077f/commands/input/input.go#L55-L69

so what's happening is that whatever you have set `core.editor` (or `GIT_EDITOR`, `EDITOR`, etc...) to (presumably, this is `subl -w`) is erroring out, so `git-bug` tries to launch it under `sh`, which fails because that isn't really something we should be doing on windows, and then you're shown that error.

what you need to do here is figure out what command you use to launch sublime text from your shell, and use _that_ as the value for `core.editor` (or `GIT_EDITOR`, `EDITOR`, etc...).

**Comment by sudoforge** [comment]
this may provide some insight for you: http://web.archive.org/web/20250711105026/https://stackoverflow.com/questions/8951275/how-can-i-make-sublime-text-the-default-editor-for-git

**Comment by olaf-kw** [comment]
Same problem as OP, using 0.10.1 on Windows11. git bug bug new opened nano, which worked fine, but after closing it, the mentioned "access denied" error occured. Needless to say, that "anyone" has access to the folders...

WebUI simply shows "error" when trying to create a new bug.

**Comment by sudoforge** [comment]
@olaf-kw what is the value of your `core.editor`?

**Comment by olaf-kw** [comment]
git config list does not show a setting for core.editor




Gesendet von Outlook für Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by kaanchan** [comment]
Additional data narrowing this on **v0.10.1 / Windows 11 (26200)**.

**It is always the empty blob.** Every failure renames onto the same hash, `e69de29bb2d1d6434b8b29ae775ad8c2e48c5391`:

```
Error: rename ...\.git\objects\pack\tmp_obj_2371093332
       ...\.git\objects\e6\9de29bb2d1d6434b8b29ae775ad8c2e48c5391: Access is denied.
```

**It only affects *fresh* repos.** In an established repo (20 bugs, created over the preceding weeks) `git-bug bug new` succeeds normally. In a newly `git init`-ed repo it fails every time. Reproduced 5+ times across two separate locations — one nested inside another git repo, one standalone — so it is not a nesting or path-depth artifact.

**Pre-creating the empty blob does not help.** Running `printf '' | git hash-object -w --stdin` first (so the loose object exists) still produces the identical rename failure, which argues against a plain rename-over-existing-read-only-file explanation.

**It is not limited to `bug new`.** `git-bug bridge pull` fails identically and imports 0 issues:

```
import error: issue creation: rename ...\objects\e6\9de29bb...: Access is denied.
imported 0 issues and 0 identities with gh bridge
```

Practical consequence: git-bug cannot be adopted in any new repository on an affected machine, and bridge testing in a throwaway sandbox is impossible — the only workable test rig is an already-initialised store.

Root cause not established here; the pattern (temp file in `objects/pack/`, rename into `objects/xx/`, `ERROR_ACCESS_DENIED`, only when the object is newly written) is consistent with the temp file still being open, or an external scanner holding it, at rename time — but that is a hypothesis, not a confirmed diagnosis.
